### PR TITLE
quayio: allow migration to skip adding manifest columns if exists (PROJQUAY-2579)

### PR DESCRIPTION
**Pull request by kleesc**
As part of the migration preparations for quayio, some new columns will
already have been added. This change allows alembic to move forward
even if these columns already exists. It also allows to skip adding a
new index to the manifestblob table if it already exists.

### CODE REVIEW

**Comment by syed** [comment]
@kleesc  can you also add a check for the `uploadedblob` table

```
Traceback (most recent call last):
  File "/app/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1246, in _execute_context
    cursor, statement, parameters, context
  File "/app/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 588, in do_execute
    cursor.execute(statement, parameters)
  File "/app/lib/python3.6/site-packages/pymysql/cursors.py", line 170, in execute
    result = self._query(query)
  File "/app/lib/python3.6/site-packages/pymysql/cursors.py", line 328, in _query
    conn.query(q)
  File "/app/lib/python3.6/site-packages/pymysql/connections.py", line 517, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "/app/lib/python3.6/site-packages/pymysql/connections.py", line 732, in _read_query_result
    result.read()
  File "/app/lib/python3.6/site-packages/pymysql/connections.py", line 1075, in read
    first_packet = self.connection._read_packet()
  File "/app/lib/python3.6/site-packages/pymysql/connections.py", line 684, in _read_packet
    packet.check_error()
  File "/app/lib/python3.6/site-packages/pymysql/protocol.py", line 220, in check_error
    err.raise_mysql_exception(self._data)
  File "/app/lib/python3.6/site-packages/pymysql/err.py", line 109, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.InternalError: (1050, "Table 'uploadedblob' already exists")
```

**Comment by syed** [comment]
`./data/migrations/versions/3383aad1e992_add_uploadedblob_table.py` releavent file

**Comment by kleesc** [comment]
@syed Done

File: data/migrations/versions/909d725887d3_add_composite_index_on_manifestblob.py [hunk 1]
```diff
@@ -11,14 +11,19 @@
...
 
 
 def upgrade(op, tables, tester):
-    op.create_index(
-        "manifestblob_repository_id_blob_id",
-        "manifestblob",
-        ["repository_id", "blob_id"],
-    )
+    inspector = Inspector.from_engine(op.get_bind())
+    manifestblob_indexes = inspector.get_indexes("manifest")
```
Comment (syed) [inline data/migrations/versions/909d725887d3_add_composite_index_on_manifestblob.py:19]: I'm getting an error 

```
sqlalchemy.exc.InternalError: (pymysql.err.InternalError) (1061, "Duplicate key name 'manifestblob_repository_id_blob_id'")
[SQL: CREATE INDEX manifestblob_repository_id_blob_id ON manifestblob (repository_id, blob_id)]
(Background on this error at: http://sqlalche.me/e/2j85)
```
The table here should be `manifestblob`

Comment (kleesc) [reply data/migrations/versions/909d725887d3_add_composite_index_on_manifestblob.py]: Fixed

**Comment by syed** [comment]
@kleesc  Migration seems to have worked successfully 

```
20:12:18 INFO  [alembic.runtime.migration] Context impl MySQLImpl.
20:12:18 INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
20:12:18 INFO  [alembic.runtime.migration] Running upgrade 88e64904d000 -> 909d725887d3, Add composite i
ndex on manifestblob
Running init script '/quay-registry/conf/init/supervisord_conf_create.sh'
Running init script '/quay-registry/conf/init/zz_boot.sh'
/quay-registry/data/secscan_model/__init__.py:28: Deprecatio
```

I will do another full test starting from py2 DB and the run the py3 upgrade.

**Comment by syed** [comment]
Ran a full test again

1. Spin up python 2 quay.io (latest iamge from quayio branch)
2. Push image to py2 quay
3. Manually run the update queries
4. Upgrade to python 3 quay
5. Validate DB migration occurs without errors


```
/quay-registry/data/secscan_model/__init__.py:30: DeprecationWarning: Call to deprecated class NoopV2SecurityScanner. (Will be replaced by a V4 API security scanner soon)
  self._legacy_model = NoopV2SecurityScanner()
22:05:06 INFO  [alembic.runtime.migration] Context impl MySQLImpl.
22:05:06 INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
22:05:06 INFO  [alembic.runtime.migration] Running upgrade 8e6a363784bb -> 04b9d2191450
22:05:06 INFO  [alembic.runtime.migration] Running upgrade 04b9d2191450 -> 3383aad1e992, Add UploadedBlob table
22:05:06 INFO  [alembic.runtime.migration] Running upgrade 3383aad1e992 -> 88e64904d000, Add new metadata columns to Manifest table
22:05:06 INFO  [alembic.runtime.migration] Running upgrade 88e64904d000 -> 909d725887d3, Add composite index on manifestblob
Running init script '/quay-registry/conf/init/supervisord_conf_create.sh'
Running init script '/quay-registry/conf/init/zz_boot.sh'
/quay-registry/data/secscan_model/__init__.py:28: DeprecationWarning: Call to deprecated class V2SecurityScanner. (Will be replaced by a V4 API security scanner soon)
  self._legacy_model = V2SecurityScanner(app, instance_keys, storage)
Failed to validate security scanner V2 configuration
```

**Review by syed — APPROVED** [review]

**Pull request merged**